### PR TITLE
[Dynamic Protocol State] Read-only interfaces of protocol state

### PR DESCRIPTION
**Pull request by durkmurder**
https://github.com/dapperlabs/flow-go/issues/5513

### Context

In this PR I've introduced read-only interfaces for accessing the protocol state. Initially we have planned to support two types of queries:
- `ByEpoch` - returns information that is static for duration of epoch, can be queried by epoch counter.
- `ByBlockID` - returns information which can change from block to block during the epoch duration, can be queried by block ID.

I have implemented only `ByBlockID` since it's unclear if `ByEpoch` is needed at this point at all. The only place where we query by-epoch is from consensus committee, which already has a logic for caching epochs on it's own. 

Why we can't simplify consensus committee to use protocol state? Replacing existing logic with protocol state is not very practical since the backbone of consensus committee is mapping views to epoch counters. This information is not available in protocol state. With protocol state we can only change from where the data will be fetched:
- from raw epoch events(which are needed to be processed either way)
- or from protocol state(after we have processed raw epoch events to obtain the counter)
To summarize, I don't see too much value in using `ByEpoch` in consensus committee, so I have omitted it's implementation at this point since it requires building an extra index in database for storing `epoch_counter` -> `protocol_state_id`.

As part of this PR I've also changed how `GlobalParams` work, in current implementation it doesn't return errors anymore since all data must be accessible at any point from any place. `GlobalParams` will be served only from protocol state when we change implementation of `protocol.Snapshot`.

### CODE REVIEW

**Comment by durkmurder** [comment]
I've accidentally merged PR for [state updater](https://github.com/onflow/flow-go/pull/4594), now it's in this PR. Sorry for such confusion. Closed PR has diff and adds changes very locally, so hopefully it isn't too bad.

File: state/protocol/protocol_state/updater_test.go [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+}
+
+// TestProcessEpochSetupHappyPath tests if processing epoch setup when invariants are not violated updates internal structures.
+// It must produce valid identity table for current and next epochs.
+// For current epoch we should have identity table with all the nodes from the current epoch + nodes from the next epoch with 0 weight.
+// For next epoch we should have identity table with all the nodes from the next epoch + nodes from the current epoch with 0 weight.
+func (s *UpdaterSuite) TestProcessEpochSetupHappyPath() {
+	setup := unittest.EpochSetupFixture(func(setup *flow.EpochSetup) {
+		setup.Counter = s.parentProtocolState.CurrentEpochSetup.Counter + 1
+		setup.Participants[0].InitialWeight = 13
```
Comment (jordanschalm) [inline state/protocol/protocol_state/updater_test.go:286]:
```suggestion
```
Is this needed?

File: state/protocol/protocol_state.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	// - (DynamicProtocolState, nil) - if there is a protocol state associated with given block ID.
+	// - (nil, storage.ErrNotFound) - if there is no protocol state associated with given block ID.
+	// - (nil, exception) - any other error should be treated as exception.
+	AtBlockID(blockID flow.Identifier) (DynamicProtocolState, error)
+	// GlobalParams returns params that are the same for all nodes in the network.
+	GlobalParams() GlobalParams
+}
+
+// StateUpdater is a dedicated interface for updating protocol state.
+// It is used by the compliance layer to update protocol state when certain events that are stored in blocks are observed.
```
Comment (jordanschalm) [inline state/protocol/protocol_state.go:70]: Based on reading the current implementation, my interpretation is that the `ParticipantState` is responsible for validating input events, and only passing valid events to `StateUpdater`, then `StateUpdater` is responsible for computing the protocol state change associated with the event. 

It would be great to:
- Document this separation of responsibilities in the interface
- Since the `StateUpdater` methods only return irrecoverable errors, document that inputs must have already passed validation (eg. `isValidExtendingEpochSetup`)

File: engine/collection/synchronization/engine.go [hunk 3]
```diff
@@ -101,7 +101,7 @@ func New(
 		return nil, fmt.Errorf("could not setup message handler")
 	}
 
-	chainID, err := state.Params().ChainID()
+	chainID := state.Params().ChainID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get chain ID: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply engine/collection/synchronization/engine.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: engine/collection/message_hub/message_hub.go [hunk 4]
```diff
@@ -156,7 +156,7 @@ func NewMessageHub(log zerolog.Logger,
 	}
 
 	// register network conduit
-	chainID, err := clusterState.Params().ChainID()
+	chainID := clusterState.Params().ChainID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get chain ID: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply engine/collection/message_hub/message_hub.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: engine/access/rpc/backend/backend_transactions.go [hunk 5]
```diff
@@ -454,7 +454,7 @@ func (b *backendTransactions) GetTransactionResultsByBlockID(
 	// user transactions in the block
 	txCount := i
 
-	sporkRootBlockHeight, err := b.state.Params().SporkRootBlockHeight()
+	sporkRootBlockHeight := b.state.Params().SporkRootBlockHeight()
 	if err != nil {
 		return nil, status.Errorf(codes.Internal, "failed to retrieve root block: %v", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply engine/access/rpc/backend/backend_transactions.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: cmd/scaffold.go [hunk 6]
```diff
@@ -1204,7 +1201,7 @@ func (fnb *FlowNodeBuilder) setRootSnapshot(rootSnapshot protocol.Snapshot) erro
 	}
 
 	fnb.RootChainID = fnb.FinalizedRootBlock.Header.ChainID
-	fnb.SporkID, err = fnb.RootSnapshot.Params().SporkID()
+	fnb.SporkID = fnb.RootSnapshot.Params().SporkID()
 	if err != nil {
 		return fmt.Errorf("failed to read spork ID: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply cmd/scaffold.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/snapshot_test.go [hunk 7]
```diff
@@ -133,17 +130,17 @@ func TestSnapshot_Params(t *testing.T) {
 		}
 		for _, snapshot := range snapshots {
 			t.Run("should be able to get chain ID from snapshot", func(t *testing.T) {
-				chainID, err := snapshot.Params().ChainID()
+				chainID := snapshot.Params().ChainID()
 				require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/snapshot_test.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/snapshot_test.go [hunk 8]
```diff
@@ -133,17 +130,17 @@ func TestSnapshot_Params(t *testing.T) {
...
 			t.Run("should be able to get spork ID from snapshot", func(t *testing.T) {
-				sporkID, err := snapshot.Params().SporkID()
+				sporkID := snapshot.Params().SporkID()
 				require.NoError(t, err)
 				assert.Equal(t, expectedSporkID, sporkID)
 			})
 			t.Run("should be able to get protocol version from snapshot", func(t *testing.T) {
-				protocolVersion, err := snapshot.Params().ProtocolVersion()
+				protocolVersion := snapshot.Params().ProtocolVersion()
 				require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/snapshot_test.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/snapshot_test.go [hunk 9]
```diff
@@ -133,17 +130,17 @@ func TestSnapshot_Params(t *testing.T) {
...
 			t.Run("should be able to get chain ID from snapshot", func(t *testing.T) {
-				chainID, err := snapshot.Params().ChainID()
+				chainID := snapshot.Params().ChainID()
 				require.NoError(t, err)
 				assert.Equal(t, expectedChainID, chainID)
 			})
 			t.Run("should be able to get spork ID from snapshot", func(t *testing.T) {
-				sporkID, err := snapshot.Params().SporkID()
+				sporkID := snapshot.Params().SporkID()
 				require.NoError(t, err)
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/snapshot_test.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/state.go [hunk 10]
```diff
@@ -544,16 +544,13 @@ func (state *State) bootstrapSporkInfo(root protocol.Snapshot) func(*badger.Txn)
...
+		err := operation.InsertSporkID(sporkID)(tx)
 		if err != nil {
 			return fmt.Errorf("could not insert spork ID: %w", err)
 		}
 
-		sporkRootBlockHeight, err := params.SporkRootBlockHeight()
+		sporkRootBlockHeight := params.SporkRootBlockHeight()
 		if err != nil {
 			return fmt.Errorf("could not get spork root block height: %w", err)
 		}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/state.go [hunk 11]
```diff
@@ -562,7 +559,7 @@ func (state *State) bootstrapSporkInfo(root protocol.Snapshot) func(*badger.Txn)
 			return fmt.Errorf("could not insert spork root block height: %w", err)
 		}
 
-		version, err := params.ProtocolVersion()
+		version := params.ProtocolVersion()
 		if err != nil {
 			return fmt.Errorf("could not get protocol version: %w", err)
 		}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/state.go [hunk 12]
```diff
@@ -571,7 +568,7 @@ func (state *State) bootstrapSporkInfo(root protocol.Snapshot) func(*badger.Txn)
 			return fmt.Errorf("could not insert protocol version: %w", err)
 		}
 
-		threshold, err := params.EpochCommitSafetyThreshold()
+		threshold := params.EpochCommitSafetyThreshold()
 		if err != nil {
 			return fmt.Errorf("could not get epoch commit safety threshold: %w", err)
 		}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/inmem/convert.go [hunk 13]
```diff
@@ -101,23 +101,23 @@ func FromParams(from protocol.GlobalParams) (*Params, error) {
 		err    error
 	)
 
-	params.ChainID, err = from.ChainID()
+	params.ChainID = from.ChainID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get chain id: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/inmem/convert.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/badger/validity.go [hunk 14]
```diff
@@ -415,7 +415,7 @@ func ValidRootSnapshotContainsEntityExpiryRange(snapshot protocol.Snapshot) erro
 		return fmt.Errorf("could not query root snapshot head: %w", err)
 	}
 
-	sporkRootBlockHeight, err := snapshot.Params().SporkRootBlockHeight()
+	sporkRootBlockHeight := snapshot.Params().SporkRootBlockHeight()
 	if err != nil {
 		return fmt.Errorf("could not query spork root block height: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/validity.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/inmem/convert.go [hunk 15]
```diff
@@ -101,23 +101,23 @@ func FromParams(from protocol.GlobalParams) (*Params, error) {
...
-	params.SporkID, err = from.SporkID()
+	params.SporkID = from.SporkID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get spork id: %w", err)
 	}
-	params.SporkRootBlockHeight, err = from.SporkRootBlockHeight()
+	params.SporkRootBlockHeight = from.SporkRootBlockHeight()
 	if err != nil {
 		return nil, fmt.Errorf("could not get spork root block height: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/inmem/convert.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/inmem/convert.go [hunk 16]
```diff
@@ -101,23 +101,23 @@ func FromParams(from protocol.GlobalParams) (*Params, error) {
...
-	params.ChainID, err = from.ChainID()
+	params.ChainID = from.ChainID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get chain id: %w", err)
 	}
-	params.SporkID, err = from.SporkID()
+	params.SporkID = from.SporkID()
 	if err != nil {
 		return nil, fmt.Errorf("could not get spork id: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/inmem/convert.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/inmem/convert.go [hunk 17]
```diff
@@ -101,23 +101,23 @@ func FromParams(from protocol.GlobalParams) (*Params, error) {
...
-	params.ProtocolVersion, err = from.ProtocolVersion()
+	params.ProtocolVersion = from.ProtocolVersion()
 	if err != nil {
 		return nil, fmt.Errorf("could not get protocol version: %w", err)
 	}
-	params.EpochCommitSafetyThreshold, err = from.EpochCommitSafetyThreshold()
+	params.EpochCommitSafetyThreshold = from.EpochCommitSafetyThreshold()
 	if err != nil {
 		return nil, fmt.Errorf("could not get protocol version: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/inmem/convert.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: state/protocol/inmem/convert.go [hunk 18]
```diff
@@ -101,23 +101,23 @@ func FromParams(from protocol.GlobalParams) (*Params, error) {
...
-	params.SporkRootBlockHeight, err = from.SporkRootBlockHeight()
+	params.SporkRootBlockHeight = from.SporkRootBlockHeight()
 	if err != nil {
 		return nil, fmt.Errorf("could not get spork root block height: %w", err)
 	}
-	params.ProtocolVersion, err = from.ProtocolVersion()
+	params.ProtocolVersion = from.ProtocolVersion()
 	if err != nil {
 		return nil, fmt.Errorf("could not get protocol version: %w", err)
 	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/inmem/convert.go]: 👉  https://github.com/onflow/flow-go/pull/4579/commits/21c621f987faf41603ad5930ede40e4f4cfb7243

File: model/flow/protocol_state.go [hunk 19]
```diff
@@ -149,6 +152,58 @@ func (e *ProtocolStateEntry) ID() Identifier {
...
+	return &ProtocolStateEntry{
+		CurrentEpochEventIDs:            e.CurrentEpochEventIDs,
+		PreviousEpochEventIDs:           e.PreviousEpochEventIDs,
+		Identities:                      e.Identities.Copy(),
+		InvalidStateTransitionAttempted: e.InvalidStateTransitionAttempted,
+		NextEpochProtocolState:          e.NextEpochProtocolState.Copy(),
+	}
+}
+
+// Copy returns a full copy of rich protocol state entry.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply model/flow/protocol_state.go:174]: 👉 https://github.com/onflow/flow-go/pull/4579/commits/8b16dadc8ae2fa9a5075d5c2279a80733bf22881

File: model/flow/protocol_state.go [hunk 20]
```diff
@@ -161,6 +216,37 @@ func (ll DynamicIdentityEntryList) Sorted(less IdentifierOrder) bool {
...
+	lenList := len(ll)
+	for i := 0; i < lenList; i++ {
+		// copy the object
+		next := *(ll[i])
+		dup = append(dup, &next)
+	}
+	return dup
+}
+
+// Sort sorts the list by the input ordering.
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply model/flow/protocol_state.go]: 👉 https://github.com/onflow/flow-go/pull/4579/commits/8b16dadc8ae2fa9a5075d5c2279a80733bf22881

File: state/protocol/inmem/initial_protocol_state.go
Comment (AlexHentschel) [inline state/protocol/inmem/initial_protocol_state.go:1]: I would be inclined to extend `RichProtocolStateEntry` such that is directly implements the interfaces `protocol.DynamicProtocolState` and `protocol.InitialProtocolState`. In other words, merge the implementation of `InitialProtocolStateAdapter` and `DynamicProtocolStateAdapter` into  `RichProtocolStateEntry`

File: state/protocol/badger/state.go [hunk 21]
```diff
@@ -667,7 +664,14 @@ func OpenState(
 }
 
 func (state *State) Params() protocol.Params {
-	return Params{state: state}
+	globalParams, err := ReadGlobalParams(state)
+	if err != nil {
+		panic("failed to read global params: " + err.Error())
+	}
```
[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go:661]: +1. Ideally, we would cache it in `badger.Params` (`state/protocol/badger/params.go`)

We could:
* use lazy loading with [`atomic.Pointer`](https://pkg.go.dev/sync/atomic#Pointer)
* add a constructor and load at initialization. Then we don't need _any_ concurrency control. 

   Presumably, the `GlobalParams` are accessed at startup of the node anyway (?) Hence, loading at construction time might be the way to go

[the remark this answers is not in the document]
Comment (AlexHentschel) [reply state/protocol/badger/state.go:661]: dang. We can't cache in cache in `badger.Params`, because `Params` depends on `badger.state`

File: state/protocol/protocol_state/updater_test.go [hunk 22]
```diff
@@ -0,0 +1,464 @@
...
+		"should have correct setup ID for next protocol state")
+	require.False(s.T(), nextEpochProtocolState.InvalidStateTransitionAttempted, "should not set invalid state transition attempted")
+	require.Nil(s.T(), nextEpochProtocolState.NextEpochProtocolState, "should not have next epoch protocol state")
+	require.Equal(s.T(), expectedNextEpochIdentityTable, nextEpochProtocolState.Identities)
+}
+
+// TestProcessEpochSetupWithSameParticipants tests that processing epoch setup with overlapping participants results in correctly
+// built updated protocol state. It should build a union of participants from current and next epoch for current and
+// next epoch protocol states respectively.
+func (s *UpdaterSuite) TestProcessEpochSetupWithSameParticipants() {
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/updater_test.go:364]: I find that just relying on length is a bit dangerous. I would suggest to extend this test (similarly to the one before) to check identities explicitly

File: state/protocol/protocol_state/updater_test.go [hunk 23]
```diff
@@ -0,0 +1,464 @@
...
+		}
+	}
+
+	// check if changes are reflected in current and next epochs
+	requireUpdatesApplied(updatedState.Identities.Lookup())
+	requireUpdatesApplied(updatedState.NextEpochProtocolState.Identities.Lookup())
+}
+
+// TestProcessEpochSetupInvariants tests if processing epoch setup when invariants are violated doesn't update internal structures.
+func (s *UpdaterSuite) TestProcessEpochSetupInvariants() {
```
Comment (AlexHentschel) [inline state/protocol/protocol_state/updater_test.go:246]: I feel we should also validate that `TransitionToNextEpoch()` errors when `InvalidStateTransitionAttempted == true`. I think currently, the implementation of `TransitionToNextEpoch()` does not really cover this case explicitly  (?)

File: model/flow/protocol_state.go [hunk 24]
```diff
@@ -154,6 +157,59 @@ func (e *ProtocolStateEntry) ID() Identifier {
...
+		CurrentEpochEventIDs:            e.CurrentEpochEventIDs,
+		PreviousEpochEventIDs:           e.PreviousEpochEventIDs,
+		Identities:                      e.Identities.Copy(),
+		InvalidStateTransitionAttempted: e.InvalidStateTransitionAttempted,
+		NextEpochProtocolState:          e.NextEpochProtocolState.Copy(),
+	}
+}
+
+// Copy returns a full copy of rich protocol state entry.
+// Embedded service events are copied by reference (not deep-copied).
```
Comment (AlexHentschel) [inline model/flow/protocol_state.go:175]:
```suggestion
//   - Embedded service events are copied by reference (not deep-copied).
//   - Identities are deep-copied, _except_ for their keys, which are copied by reference.
```

File: model/flow/protocol_state.go [hunk 25]
```diff
@@ -154,6 +157,59 @@ func (e *ProtocolStateEntry) ID() Identifier {
 	return MakeID(body)
 }
 
+// Copy returns a full copy of the entry.
```
Comment (AlexHentschel) [inline model/flow/protocol_state.go:160]:
```suggestion
// Copy returns a full copy of the entry.
// Embedded Identities are deep-copied, _except_ for their keys, which are copied by reference.
```

File: model/flow/protocol_state.go [hunk 26]
```diff
@@ -166,6 +222,37 @@ func (ll DynamicIdentityEntryList) Sorted(less IdentifierOrder) bool {
...
+func (ll DynamicIdentityEntryList) ByNodeID(nodeID Identifier) (*DynamicIdentityEntry, bool) {
+	for _, identity := range ll {
+		if identity.NodeID == nodeID {
+			return identity, true
+		}
+	}
+	return nil, false
+}
+
+func (ll DynamicIdentityEntryList) Copy() DynamicIdentityEntryList {
```
Comment (AlexHentschel) [inline model/flow/protocol_state.go:235]:
```suggestion
// Copy returns a copy of the DynamicIdentityEntryList. The resulting slice uses
// a different backing array, meaning appends and insert operations on either slice
// are guaranteed to only affect that slice.
//
// Copy should be used when modifying an existing identity list by either
// appending new elements, re-ordering, or inserting new elements in an
// existing index.
//
// CAUTION:
// All Identity fields are deep-copied, _except_ for their keys, which
// are copied by reference.
func (ll DynamicIdentityEntryList) Copy() DynamicIdentityEntryList {
```

File: model/flow/protocol_state.go [hunk 27]
```diff
@@ -166,6 +222,37 @@ func (ll DynamicIdentityEntryList) Sorted(less IdentifierOrder) bool {
...
+	lenList := len(ll)
+	for i := 0; i < lenList; i++ {
+		// copy the object
+		next := *(ll[i])
+		dup = append(dup, &next)
+	}
+	return dup
+}
+
+// Sort sorts the list by the input ordering. Returns a new, sorted list without modifying the input.
```
Comment (AlexHentschel) [inline model/flow/protocol_state.go:247]:
```suggestion
// Sort sorts the list by the input ordering. Returns a new, sorted list without modifying the input.
// CAUTION:
// All Identity fields are deep-copied, _except_ for their keys, which are copied by reference.
```

File: model/flow/protocol_state_test.go [hunk 28]
```diff
@@ -94,3 +94,29 @@ func TestNewRichProtocolStateEntry(t *testing.T) {
 		assert.Equal(t, expectedIdentities, richEntry.NextEpochProtocolState.Identities, "should be equal to next epoch setup participants + current epoch setup participants")
 	})
 }
+
+// TestProtocolStateEntry_Copy tests if the copy method returns a deep copy of the entry. All changes to cpy shouldn't affect the original entry.
```
Comment (AlexHentschel) [inline model/flow/protocol_state_test.go:98]:
```suggestion
// TestProtocolStateEntry_Copy tests if the copy method returns a deep copy of the entry.
// All changes to copy shouldn't affect the original entry -- except for key changes.
```

**Review by AlexHentschel — APPROVED** [review]

File: state/protocol/protocol_state/updater.go [hunk 29]
```diff
@@ -0,0 +1,264 @@
...
+		return fmt.Errorf("expected to find identity for current epoch, but (%v) not found", updated.NodeID)
+	}
+	currentEpochIdentity.Dynamic = newData
+
+	if u.state.NextEpochProtocolState != nil {
+		nextEpochIdentity, found := u.nextEpochIdentitiesLookup[updated.NodeID]
+		if found {
+			nextEpochIdentity.Dynamic = newData
+		}
+	}
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state/updater.go]: This suggestion isn't compatible with the [design proposal](https://www.notion.so/dapperlabs/Proposal-Detailed-specification-for-framework-functionality-to-support-future-slashing-37b7d127c75d4834becdf3d38cb2731d?pvs=4) we've agreed upon.

File: state/protocol/protocol_state/updater.go [hunk 30]
```diff
@@ -0,0 +1,264 @@
...
+	return nil
+}
+
+// SetInvalidStateTransitionAttempted sets a flag indicating that invalid state transition was attempted.
+func (u *Updater) SetInvalidStateTransitionAttempted() {
+	u.state.InvalidStateTransitionAttempted = true
+	if u.state.NextEpochProtocolState != nil {
+		u.state.NextEpochProtocolState.InvalidStateTransitionAttempted = true
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state/updater.go:204]: Captured in https://github.com/onflow/flow-go/issues/4649

File: state/protocol/protocol_state.go (hunk 2 above)
Comment (jordanschalm) [reply state/protocol/protocol_state.go:70]: Added this to doc.

File: state/protocol/protocol_state.go [hunk 31]
```diff
@@ -0,0 +1,106 @@
+package protocol
+
+import (
+	"github.com/onflow/flow-go/model/flow"
+)
+
+// InitialProtocolState returns constant data for given epoch.
+// This interface can be only obtained for epochs that have progressed to epoch commit event.
+type InitialProtocolState interface {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state.go:9]: Captured in https://github.com/onflow/flow-go/issues/4649

[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state.go:9]: Captured in https://github.com/onflow/flow-go/issues/4649

File: state/protocol/protocol_state.go [hunk 32]
```diff
@@ -0,0 +1,106 @@
...
+	DKG() (DKG, error)
+	// Entry Returns low-level protocol state entry that was used to initialize this object.
+	// It shouldn't be used by high-level logic, it is useful for some cases such as bootstrapping.
+	// Prefer using other methods to access protocol state.
+	Entry() *flow.RichProtocolStateEntry
+}
+
+// DynamicProtocolState extends the InitialProtocolState with data that can change from block to block.
+// It can be used to access the identity table at given block.
+type DynamicProtocolState interface {
```
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state.go:30]: Captured in https://github.com/onflow/flow-go/issues/4649

File: state/protocol/protocol_state/updater.go
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/protocol_state/updater.go:1]: Added a TODO for 1st point in `0d961bc047`

File: state/protocol/badger/state.go (hunk 21 above)
[the remark this answers is not in the document]
Comment (jordanschalm) [reply state/protocol/badger/state.go:661]: Captured as low-prio item in https://github.com/onflow/flow-go/issues/4649

**Review by jordanschalm — APPROVED** [review]

File: state/protocol/inmem/initial_protocol_state.go
Comment (durkmurder) [reply state/protocol/inmem/initial_protocol_state.go:1]: Mentioned it in https://github.com/onflow/flow-go/issues/4649

File: state/protocol/protocol_state.go [hunk 33]
```diff
@@ -0,0 +1,106 @@
...
+	EpochSetup() *flow.EpochSetup
+	// EpochCommit returns original epoch commit event that was used to update the protocol state.
+	EpochCommit() *flow.EpochCommit
+	// DKG returns information about DKG that was obtained from EpochCommit event.
+	// No errors are expected during normal operations.
+	DKG() (DKG, error)
+	// Entry Returns low-level protocol state entry that was used to initialize this object.
+	// It shouldn't be used by high-level logic, it is useful for some cases such as bootstrapping.
+	// Prefer using other methods to access protocol state.
+	Entry() *flow.RichProtocolStateEntry
```
[the remark this answers is not in the document]
Comment (durkmurder) [reply state/protocol/protocol_state.go:25]: Added a point [here](https://github.com/onflow/flow-go/issues/4649). It's dangerous but has it's own usages. I would suggest to implement everything and then try to get rid of it when we see how everything comes together.

File: state/protocol/protocol_state/updater_test.go (hunk 22 above)
Comment (durkmurder) [reply state/protocol/protocol_state/updater_test.go:376]: https://github.com/onflow/flow-go/pull/4579/commits/d222fd0a645ebd03a8c27df26857cee47c354689

File: state/protocol/protocol_state/updater_test.go (hunk 23 above)
Comment (durkmurder) [reply state/protocol/protocol_state/updater_test.go:258]: https://github.com/onflow/flow-go/pull/4579/commits/d222fd0a645ebd03a8c27df26857cee47c354689
